### PR TITLE
Working prototype

### PR DESCRIPTION
**Pull request by JonathanFenske**
modified featurization to create memory maps, modified train for memory maps and a generator

### CODE REVIEW

File: src/featurization.py [hunk 1]
```diff
@@ -21,20 +21,34 @@
...
         os.makedirs(path)
-    except OSError as exc:  # Python >2.5
-        if exc.errno == errno.EEXIST and os.path.isdir(path):
-            pass
-        else:
-            raise
 
+def data2memmap(file,mmap,no_of_reviews,maxlen):
+    mmap = os.path.join('data/features/',mmap)
+    data = np.memmap(mmap, dtype='float', mode='w+', shape=(no_of_reviews, maxlen, 80))
```
Comment (DennisHamann) [inline src/featurization.py:37]: 80 sollte als Parameter übergeben werden.

Comment (DennisHamann) [inline src/featurization.py:37]: Die shapes bitte noch in data/features/trainSetShape.txt speichern. memmaps wissen die eigene Größe nicht.

File: src/featurization.py [hunk 2]
```diff
@@ -146,28 +159,57 @@ def getAvgFeatureVecs(reviews, model, num_features):
...
+    test_path = os.path.join(output, f'{num_features}testDataVec.csv')
+    with open(test_path, "w") as csv_file:
+        writer = csv.writer(csv_file, delimiter=',')
+        for line in testDataVecs:
+            writer.writerow(line)
+
+    # initializing the memory maps
+    no_of_reviews_train = 40000
+    no_of_reviews_test = 5000
+    maxlen = 1000
```
Comment (DennisHamann) [inline src/featurization.py:206]: wieso maxlen=1000? Kann das nicht dynamisch bestimmt werden?

File: src/featurization.py [hunk 3]
```diff
@@ -21,20 +21,34 @@
...
+def data2memmap(file,mmap,no_of_reviews,maxlen):
+    mmap = os.path.join('data/features/',mmap)
+    data = np.memmap(mmap, dtype='float', mode='w+', shape=(no_of_reviews, maxlen, 80))
+    print('mmap=0')
+    print('start iterator')
+    for (idx, row) in enumerate(file):
+        review_length = len(row)
+        review_length = min(review_length,maxlen)
+        review = list()
+        print(idx)
```
Comment (DennisHamann) [inline src/featurization.py:44]: dieses Statement ist überflüssig

File: src/featurization.py [hunk 4]
```diff
@@ -21,20 +21,34 @@
...
-            pass
-        else:
-            raise
 
+def data2memmap(file,mmap,no_of_reviews,maxlen):
+    mmap = os.path.join('data/features/',mmap)
+    data = np.memmap(mmap, dtype='float', mode='w+', shape=(no_of_reviews, maxlen, 80))
+    print('mmap=0')
+    print('start iterator')
+    for (idx, row) in enumerate(file):
```
Comment (DennisHamann) [inline src/featurization.py:40]: für eine Prozessübersicht kann tqdm(https://tqdm.github.io/) verwendet werden.

File: src/train.py [hunk 5]
```diff
@@ -42,38 +42,42 @@ def get_model(input_shape,output_shape):
     model.compile('adam', 'binary_crossentropy', metrics=['accuracy'])
     return model
 
-def get_data(no_of_reviews,maxlen):
-    trainX = np.memmap('features/trainmapX', dtype='float', mode='r',shape=(no_of_reviews,maxlen,10))
-    testX = np.memmap('features/testmapX', dtype='float', mode='r',shape=(no_of_reviews,maxlen,10))
+def get_data(no_of_reviews_train,no_of_reviews_test,maxlen):
+    trainX = np.memmap('data/features/trainmapX', dtype='float', mode='r',shape=(no_of_reviews_train,maxlen,80))
```
Comment (DennisHamann) [inline src/train.py:46]: 80 als parameter übergeben

File: src/train.py [hunk 6]
```diff
@@ -42,38 +42,42 @@ def get_model(input_shape,output_shape):
...
-    trainY = np.array(trainYfile)
-    testY = np.array(testYfile)
+    trainYfile = open('data/features/80trainDatalabel.pickle', 'rb')
+    testYfile = open('data/features/80testDatalabel.pickle', 'rb')
+    trainY = pickle.load(trainYfile)
+    testY = pickle.load(testYfile)
     return trainX, testX, trainY, testY
 
+#python src/train.py data/features data/models
 def data2memmap(file,mmap,no_of_reviews,maxlen):
```
Comment (DennisHamann) [inline src/train.py:57]: diese funktion ist schon in featurization.py und ist hier nicht mehr nötig

File: src/train.py [hunk 7]
```diff
@@ -95,23 +99,25 @@ def generator(X,Y,batch_size):
         f.write('Output')
```
Comment (DennisHamann) [inline src/train.py:99]: wieso wird hier Output in die .h5 Datei geschrieben?

File: src/train.py [hunk 8]
```diff
@@ -95,23 +99,25 @@ def generator(X,Y,batch_size):
...
 
     #needs to be changed
-    no_of_reviews = 4
-    maxlen = 42
+    no_of_reviews_train = 40000
```
Comment (DennisHamann) [inline src/train.py:102]: Parameter bitte automatisch auslesen

File: src/train.py [hunk 9]
```diff
@@ -27,7 +27,7 @@ def mkdir_p(path):
             raise
 
 def train_model(model,trainX,trainY,testX,testY):
-    model.fit(generator(trainX,trainY,1), nb_epoch=10, sample_per_epoch=2, verbose=1, validation_data=generator(testX,testY,1))
+    model.fit(generator(trainX,trainY,25), epochs=5, steps_per_epoch=100, verbose=1, validation_data=generator(testX,testY,25), validation_steps=50)
```
Comment (DennisHamann) [inline src/train.py:30]: steps per epoch ist über die shape der Daten und der Batchsize abhängig.

File: src/train.py [hunk 10]
```diff
@@ -42,38 +42,42 @@ def get_model(input_shape,output_shape):
...
     data.flush()
 
 def generator(X,Y,batch_size):
-    batchX = np.zeros((batch_size,X.shape[1],10))
+    batchX = np.zeros((batch_size,X.shape[1],80))
     batchY = np.zeros((batch_size,1))
     while True:
         for i in range(batch_size):
-            idx = random.choice(X.shape[0],1)
+            idx = random.choice(range(X.shape[0]-1))
```
Comment (DennisHamann) [inline src/train.py:80]: mit random.choice werden die Trainingsbeispiele unterschiedlich oft zurückgegeben. Hier sollte eine Permutation (z.B. shuffle() verwendet werden.
